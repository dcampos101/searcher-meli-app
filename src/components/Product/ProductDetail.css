/* estilo detalle*/
.img-shipping{
    margin: 0 5px 0 0;
  }
  .img-shipping-truck{
    width:20px;
    height:20px;
  }
  /*--------*/
  .section-product {
    max-width: 1000px;
    background-color: rgb(251, 249, 249);
    margin: 10px auto;
    padding: 10px 20px;
    min-height: 600px;
    box-shadow: 1px 2px 0 rgba(0.25, 0.25, 0.25, 0.25);
    border-radius: 4px;
    display: flex;
    flex-direction: column;
    user-select: none;
  
  
  }
  .main-product {
    display: flex;
    justify-content: center;
    flex: 50% 0 0;
  }
  .features-product {
    flex: 1;
    font-size: 14px;
  }
  .content-product {
    display: flex;
  
  }
  .content-product {
    display: flex;
   
  }
  .img-container {
    max-width: 240px;
    max-height: 330px;
  
  }
  
  .img-container img {
    width: 100%;
  }
  
  .details-product {
    flex: 35%;
    font-family: Proxima Nova,-apple-system,Helvetica Neue,Helvetica,Roboto,Arial,sans-serif;
  }
  
  .item-details {
    margin-top: 16px;
    font-size: 14px
  }
  .item-condition {
    font-size: 1.0rem;
    font-weight: 500;
    color: #00a650;
    margin: 0 0.5rem 0 0;
  }
  
  .item-shipping{
    display:flex;
    justify-content: flex-start;
    align-items: center;
    margin-top: 15px;
  }
  
  .item-shipping p{
    margin: 0 45px 0 0; 
    padding: 0 120px 0 0;
   }
   .tem-shipping img-shipping{
    margin: 0 5px 0 0;
  }
  .tem-shipping item-free_shipping {
    font-weight: 400;
    line-height: 1.3;
    color: #00a650;
    font-size: 16px;
    text-align: start;
  }
  
  .tem-shipping img-shipping-truck{
    width:20px;
    height:20px;
  }
  
  .features-text {
    font-size: 18px;
    font-weight: 400;
    color: rgba(0, 0, 0, 0.9);
    line-height: 1.25;
  }
  .text_price {
    font-size: 1.5rem;
    margin: 0;
  }
  .features-grid {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(calc(min(100%,100px),1fr)));
    grid-gap: 30px 50px;
  }
  .button2 {
    background-color: #008CBA;
  }
  .button {
    background-color: #3483fa;  
    border: none;
    color: white;
    padding: 12px 50px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 14px;
    border-radius: 6px;
    align-items: left;
    cursor: pointer;
    font-family: Proxima Nova,-apple-system,Helvetica Neue,Helvetica,Roboto,Arial,sans-serif;
    font-size: 14px;
    font-weight: 400;
    height: 40px;
}