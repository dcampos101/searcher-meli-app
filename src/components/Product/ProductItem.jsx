import { BsFillLightningFill } from "react-icons/bs";
import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom"
import PropTypes from 'prop-types'

export const ProductItem = ({ title, url, id, prices, shipping }) => {
  const [totalReviews, setTotalReviews] = useState(0);
  const [ratingAverage, setRatingAverage] = useState(0);

function discount_rate(high_price, low_price) {
  const discount = (high_price - low_price) / high_price;
  return Math.floor(discount * 100);
}
    return (
      <Link to={id} className="card ">
        <article>  
            <img src={url} alt={title} />
            <section>
              <h3>{title}</h3>
              <section className="details_product">
                <div>
                  {prices.prices[0].amount !==
                    prices.prices[prices.prices.length - 1].amount && (
                    <span className="max_price">
                      ${new Intl.NumberFormat().format(prices.prices[0].amount)}
                    </span>
                  )}
                  <div className="price_container">
                    <p className="text_price">
                      $
                      {new Intl.NumberFormat().format(
                        prices.prices[prices.prices.length - 1].amount
                      )}
                    </p>
                    {prices.prices[0].amount !==
                      prices.prices[prices.prices.length - 1].amount && (
                      <span>
                        {discount_rate(
                          prices.prices[0].amount,
                          prices.prices[prices.prices.length - 1].amount
                        )}
                        % OFF
                      </span>
                    )}
                  </div>
                  <div className="div_shipping_container">
                    {shipping.free_shipping && (
                      <p className="text_envio">Envio gratis</p>
                    )}
                    {shipping.logistic_type === "fulfillment" && (
                      <div>
                        <BsFillLightningFill size={15} color="#00a650" />
                        <span className="full">FULL</span>
                      </div>
                    )}
                  </div>
                </div>
                <div className="reviews_container">
                  {ratingAverage === 0 ? null : render_stars(ratingAverage)}
                  {ratingAverage === 0 ? null : (
                    <span className="total_reviews">{totalReviews}</span>
                  )}
                </div>
              </section>
            </section>
        </article>
      </Link>
    )
  }
  
  ProductItem.propTypes = {
    title: PropTypes.string.isRequired, 
    url: PropTypes.string.isRequired, 
    id: PropTypes.string.isRequired
  }