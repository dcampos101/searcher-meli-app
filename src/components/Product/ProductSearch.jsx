import { useState } from 'react';
import { BsSearch } from 'react-icons/bs';
import PropTypes from 'prop-types';

export const ProductSearch = ({ onNewCategory }) => {

    const [ inputValue, setInputValue ] = useState('');

    const onInputChange = ({ target }) => {
        setInputValue( target.value );
    }

    const onSubmit = ( event ) => {
        event.preventDefault();
        if( inputValue.trim().length <= 1) return;

        setInputValue('');
        onNewCategory( inputValue.trim() );
    }

    return (
        <form onSubmit={ onSubmit } aria-label='form'>
            <input 
                type="text"
                placeholder="Buscar productos, marcas y más…"
                maxLength={ 100 }
                value={ inputValue }
                onChange={ onInputChange }
            />
            <BsSearch className="search_icon" color="#999" size={20} />
        </form>
    )
}

ProductSearch.propTypes = {
    onNewCategory: PropTypes.func.isRequired
}