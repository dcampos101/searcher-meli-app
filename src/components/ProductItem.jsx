import {
  BsFillLightningFill,
  BsStar,
  BsStarFill,
  BsStarHalf,
} from "react-icons/bs";
import React, { useEffect, useState } from "react";


export const ProductItem = ({ title, url, id, condition, price, prices, shipping }) => {
  const [totalReviews, setTotalReviews] = useState(0);
  const [ratingAverage, setRatingAverage] = useState(0);

//          <img src={ url } alt={ title } />
//<p>{ title }</p>
/*
        id: product.id,
        title: product.title,
        condition: product.condition,
        price: product.price,
        //shipping: shipping.free_shipping,
        url: product.thumbnail
*/
function discount_rate(high_price, low_price) {
  const discount = (high_price - low_price) / high_price;
  return Math.floor(discount * 100);
}
    return (
        <article>
        <img src={url} alt={title} />
        <section>
          <h3>{title}</h3>
          <section className="details_product">
            <div>
              {prices.prices[0].amount !==
                prices.prices[prices.prices.length - 1].amount && (
                <span className="max_price">
                  ${new Intl.NumberFormat().format(prices.prices[0].amount)}
                </span>
              )}
              <div className="price_container">
                <p className="text_price">
                  $
                  {new Intl.NumberFormat().format(
                    prices.prices[prices.prices.length - 1].amount
                  )}
                </p>
                {prices.prices[0].amount !==
                  prices.prices[prices.prices.length - 1].amount && (
                  <span>
                    {discount_rate(
                      prices.prices[0].amount,
                      prices.prices[prices.prices.length - 1].amount
                    )}
                    % OFF
                  </span>
                )}
              </div>
              <div className="div_shipping_container">
                {shipping.free_shipping && (
                  <p className="text_envio">Envio gratis</p>
                )}
                {shipping.logistic_type === "fulfillment" && (
                  <div>
                    <BsFillLightningFill size={15} color="#00a650" />
                    <span className="full">FULL</span>
                  </div>
                )}
              </div>
            </div>
            <div className="reviews_container">
              {ratingAverage === 0 ? null : render_stars(ratingAverage)}
              {ratingAverage === 0 ? null : (
                <span className="total_reviews">{totalReviews}</span>
              )}
            </div>
          </section>
        </section>
      </article>
    )
  }
  