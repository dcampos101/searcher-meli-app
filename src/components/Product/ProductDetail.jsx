import { useParams } from "react-router-dom";
import React,{useState,useEffect} from 'react'
import Spinner from "../Spinner/Spinner";
import Feature from "../Feature/Feature";
import ImgDet from "../ImgDet/ImgDet";
import { Carousel } from 'react-responsive-carousel';
import './ProductDetail.css';
import "react-responsive-carousel/lib/styles/carousel.min.css";

export const ProductDetail =  () => {
  const { id } = useParams();
    
  const [details, setDetails] = useState([]);
  const [loading, setLoading] = useState(true);
  async function getProduct() {
      const data = await fetch(
          `https://api.mercadolibre.com/items/${ id }`
        );

        const dataJson = await data.json();
        setDetails(dataJson);
        setLoading(false);
  }

  useEffect(() => {
       getProduct();
  }, []);
 
      return (
        <>
          {loading ? (
            <Spinner />
          ) : (
            <section className="section-product">
              <div className="main-product">
                <div className="content-product">
                  <figure className="img-container">
                  <Carousel className="slider-wrapper.axis-vertical">
                    {
                      details.pictures.map((value, index) => {
                        return <img key="{index}" className="img" src={value.url}></img>
                      })
                    }
                  </Carousel>
                  </figure>
                </div>
                <div className="details-product">
                  <div className="item-details">
                    <span className="item-condition">
                      {details.condition === "new" ? "Nuevo" : ""}
                    </span>
                    <h1>{details.title}</h1>
                    <p className="text_price">
                      $
                      {new Intl.NumberFormat().format(
                        details.price
                      )}
                    </p>
                    <div className="item-shipping">
                      <figure className="img-shipping">
                        <div className="img-shipping-truck">
                          <ImgDet />
                        </div>
                      </figure>
                      <p className="item-free_shipping">
                        {details.shipping.free_shipping
                          ? "Envío gratis a todo el país "
                          : ""}
                      </p>
                      <button className="button">Comprar ahora</button>
                    </div>
                    <div className="item-shipping"> </div>
                  </div>
                  <div className="features-product">
                    <h2 className="features-text">
                      Características de {details.title}
                    </h2>
                    <div className="features-grid">
                      {details.attributes.slice(0, 12).map((attr, key) => {
                        return (
                          <Feature
                            name={attr.name}
                            value={attr.value_name}
                            key={key}
                          />
                        );
                      })}
                    </div>
                  </div>
                </div>
              </div>
            </section>
          )}
        </>
      );
}

export default ProductDetail;