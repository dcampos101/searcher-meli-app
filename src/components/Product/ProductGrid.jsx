import { ProductItem } from './ProductItem';
import { useFetchProduct } from '../../hooks/useFetchProduct';
import PropTypes from 'prop-types';

export const ProductGrid = ({ category }) => {

    const { products, isLoading } = useFetchProduct( category );
    
    return (
        <>
        <div> 
            <div className="card-grid" >
                {
                    products.map( ( prod ) => (
                        <ProductItem
                            key={ prod.id } 
                            { ...prod }
                        />
                    ))
                }
            </div>
        </div>
        </>
    )
}

ProductGrid.propTypes = {
    category: PropTypes.string.isRequired
}