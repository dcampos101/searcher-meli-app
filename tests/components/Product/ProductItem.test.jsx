import { render, screen } from "@testing-library/react"
import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import { ProductItem } from "../../../src/components/Product/ProductItem"

describe('Prueba en ProductItem', () =>{

    const title = 'Samsung Galaxy A33 5g 128 Gb Awesome White 6 Gb Ram';
    const url = 'http://http2.mlstatic.com/D_963008-MLA50999956957_082022-I.jpg';
    const id = 'MLA1137723021';
    const prices = {
        prices: {
            0: {
                amount: 99999,
                conditions: {context_restrictions: Array(0), start_time: null, end_time: null, eligible: true},
                currency_id: "ARS",
                exchange_rate_context: "DEFAULT",
                id: "11",
                last_updated: "2022-08-01T14:18:09Z",
                metadata: {},
                regular_amount: null,
                type: "standard"
            },
            1:{
                amount: 84999,
                conditions: {context_restrictions: Array(1), start_time: '2022-09-18T03:00:00Z', end_time: '2022-09-19T02:59:59Z', eligible: true},
                currency_id: "ARS",
                exchange_rate_context: "DEFAULT",
                id: "26",
                last_updated: "2022-09-16T20:04:45Z",
                metadata: {promotion_id: 'OFFER-MLA1137723021-412721015', promotion_type: 'deal_of_the_day'},
                regular_amount: 99999,
                type: "promotion"
            },
            length: 2
        }
    };
    const shipping = {
        free_shipping: true,
        logistic_type: "fulfillment",
        mode: "me2",
        store_pick_up: false
    };

    test('Debe hacer match con el snapshot', () =>{
      const { container } =  render (
            <Router>
                <ProductItem title={ title } url ={ url } id={ id } prices={ prices} shipping={ shipping } />)
            </Router>
        );
        expect(container).toMatchSnapshot();

    });

    test('Debe mostrar la imagen con sus propiedades', () =>{
        render (
            <Router>
                <ProductItem title={ title } url ={ url } id={ id } prices={ prices} shipping={ shipping } />
            </Router>
        );
        const { src, alt } = screen.getByRole('img');
        expect( src ).toBe(url);
        expect( alt ).toBe(title);
    });

    test('Debe mostrar titulo del componente', () =>{
        render (
            <Router>
                <ProductItem title={ title } url ={ url } id={ id } prices={ prices} shipping={ shipping } />)
            </Router>
        );
        
        expect( screen.getByText(title)).toBeTruthy;
    });

})