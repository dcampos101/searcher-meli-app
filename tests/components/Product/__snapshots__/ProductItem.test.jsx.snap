// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en ProductItem Debe hacer match con el snapshot 1`] = `
<div>
  <a
    class="card "
    href="/MLA1137723021"
  >
    <article>
      <img
        alt="Samsung Galaxy A33 5g 128 Gb Awesome White 6 Gb Ram"
        src="http://http2.mlstatic.com/D_963008-MLA50999956957_082022-I.jpg"
      />
      <section>
        <h3>
          Samsung Galaxy A33 5g 128 Gb Awesome White 6 Gb Ram
        </h3>
        <section
          class="details_product"
        >
          <div>
            <span
              class="max_price"
            >
              $
              99.999
            </span>
            <div
              class="price_container"
            >
              <p
                class="text_price"
              >
                $
                84.999
              </p>
              <span>
                15
                % OFF
              </span>
            </div>
            <div
              class="div_shipping_container"
            >
              <p
                class="text_envio"
              >
                Envio gratis
              </p>
              <div>
                <svg
                  color="#00a650"
                  fill="currentColor"
                  height="15"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: rgb(0, 166, 80);"
                  viewBox="0 0 16 16"
                  width="15"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.52.359A.5.5 0 0 1 6 0h4a.5.5 0 0 1 .474.658L8.694 6H12.5a.5.5 0 0 1 .395.807l-7 9a.5.5 0 0 1-.873-.454L6.823 9.5H3.5a.5.5 0 0 1-.48-.641l2.5-8.5z"
                  />
                </svg>
                <span
                  class="full"
                >
                  FULL
                </span>
              </div>
            </div>
          </div>
          <div
            class="reviews_container"
          />
        </section>
      </section>
    </article>
  </a>
  )
</div>
`;
